### PR TITLE
XRT-493 xbutil --new validate

### PR DESCRIPTION
**Pull request by AShivangi**
### Work Done
- Added most of the test cases for xbutil validate
- Progress bar for verify and bandwidth tests
- --quick option to exit after verify
- Added --json support
- Added colors for passed, failed and passed with warning for each test
- Works for SSV2 platforms
- Compiles on windows

### Work to be addressed by the upcoming PRs
- Add remaining tests: m2mtest, p2ptest
- Fix platform info for ssv2 platforms (currently this info is being fetched from feature rom)
- Codeshare between firmware_image.cpp and device.cpp
- Remove Linux specific directives from validate, i.e., get this working on windows

### sample outputs
```
$xbutil --new validate
-----------------------------------------------------
 Invoking next generation of the xbutil application
-----------------------------------------------------
Starting validation for 1 devices

Validate device[0000:b3:00.1]
Platform            : xilinx_u200_xdma_201830_2
SC Version          : 4.2.0
Platform ID         : 0x5d1211e8

1/9 Test #1         : Kernel version
    Description     : Check if kernel version is supported by XRT
    [PASSED]
------------------------------------------------------------------
2/9 Test #2         : Aux connection
    Description     : Check if auxiliary power is connected
    Warning         : Aux power is not connected
    Warning         : Device is not stable for heavy acceleration tasks
    [PASSED WITH WARNINGS]
------------------------------------------------------------------
3/9 Test #3         : PCIE link
    Description     : Check if PCIE link is active
    [PASSED]
------------------------------------------------------------------
4/9 Test #4         : SC version
    Description     : Check if SC firmware is up-to-date
    [PASSED]
------------------------------------------------------------------
5/9 Test #5         : Verify kernel
    Description     : Run 'Hello World' kernel test
    [PASSED]
------------------------------------------------------------------
6/9 Test #6         : DMA
    Description     : Run dma test
    Details         : Host -> PCIe -> FPGA write bandwidth = 9225.017455 MB/s
    Details         : Host <- PCIe <- FPGA read bandwidth = 11706.670173 MB/s
    [PASSED]
------------------------------------------------------------------
7/9 Test #7         : Bandwidth kernel
    Description     : Run 'bandwidth kernel' and check the throughput
    Details         : Maximum throughput: 47798 MB/s
    [PASSED]
------------------------------------------------------------------
8/9 Test #8         : Peer to peer bar
    Description     : Run P2P test
    [SKIPPED]
------------------------------------------------------------------
9/9 Test #9         : Memory to memory DMA
    Description     : Run M2M test
    [SKIPPED]
------------------------------------------------------------------
Validation completed, but with warnings

$xbutil --new validate -qj
-----------------------------------------------------
 Invoking next generation of the xbutil application
-----------------------------------------------------
{
    "logical_devices": [
        {
            "device_id": "0000:b3:00.1",
            "platform": "xilinx_u200_xdma_201830_2",
            "sc_version": "4.2.0",
            "platform_id": "0x5d1211e8",
            "tests": [
                {
                    "id": "1",
                    "name": "Kernel version",
                    "description": "Check if kernel version is supported by XRT",
                    "status": "passed"
                },
                {
                    "id": "2",
                    "name": "Aux connection",
                    "description": "Check if auxiliary power is connected",
                    "log": [
                        {
                            "Warning": "Aux power is not connected"
                        },
                        {
                            "Warning": "Device is not stable for heavy acceleration tasks"
                        }
                    ],
                    "status": "passed"
                },
                {
                    "id": "3",
                    "name": "PCIE link",
                    "description": "Check if PCIE link is active",
                    "status": "passed"
                },
                {
                    "id": "4",
                    "name": "SC version",
                    "description": "Check if SC firmware is up-to-date",
                    "status": "passed"
                },
                {
                    "id": "5",
                    "name": "Verify kernel",
                    "description": "Run 'Hello World' kernel test",
                    "status": "passed"
                }
            ]
        }
    ]
}
```

### CODE REVIEW

File: src/runtime_src/core/common/device.cpp [hunk 1]
```diff
@@ -102,6 +119,89 @@ get_axlf_section_or_error(axlf_section_kind section, const uuid& xclbin_id) cons
   throw std::runtime_error("no such xclbin section");
 }
 
+std::pair<const char*, size_t>
+device::
+get_axlf_section(const std::string& filename, axlf_section_kind kind) const
+{
```
Comment (stsoe) [inline src/runtime_src/core/common/device.cpp:125]: Who will use this API?  And why is an member function of xrt_core::device, it doesn't seem to be related to device at all?

File: src/runtime_src/core/common/device.cpp [hunk 2]
```diff
@@ -102,6 +119,89 @@ get_axlf_section_or_error(axlf_section_kind section, const uuid& xclbin_id) cons
...
+  in.seekg(section->m_sectionOffset);
+  in.read(buf, section->m_sectionSize);
+
+  return std::make_pair(buf, section->m_sectionSize);
+}
+
+std::vector<std::string>
+device::
+get_uuids(const void *dtbuf) const
+{
```
Comment (stsoe) [inline src/runtime_src/core/common/device.cpp:165]: This also doesn't seem device related?

File: src/runtime_src/core/common/device.h [hunk 3]
```diff
@@ -221,6 +221,26 @@ class device : public ishim
...
+  /**
+   * get_uuids() - Get UUIDs from the axlf section
+   *
+   * dtbuf: axlf section to be parsed
+   *
+   * Return: list of UUIDs
+   */
+  std::vector<std::string>
+  get_uuids(const void *dtbuf) const;
+
```
Comment (stsoe) [inline src/runtime_src/core/common/device.h:243]: Should not be part of xrt::core_device unless it using the device object some how, I couldn't find that.

File: src/runtime_src/core/common/device.h [hunk 4]
```diff
@@ -306,6 +326,19 @@ class device : public ishim
...
+  uint32_t totalsize;
+  uint32_t off_dt_struct;
+  uint32_t off_dt_strings;
+  uint32_t off_mem_rsvmap;
+  uint32_t version;
+  uint32_t last_comp_version;
+  uint32_t boot_cpuid_phys;
+  uint32_t size_dt_strings;
+  uint32_t size_dt_struct;
+  };
```
Comment (stsoe) [inline src/runtime_src/core/common/device.h:341]: Does it belong in the xrt_core::device class?

File: src/runtime_src/core/common/device.cpp (hunk 1 above)
Comment (AShivangi) [reply src/runtime_src/core/common/device.cpp:125]: Yes, you're right. This and the next function aren't related to the device at all. My bad. Both of these functions are used by "SubCmdValidate.cpp" and "xbmgmt2/flash/firmware_image.cpp". Right now I created a copy and put it in here (firmware_image has its own copy of these functions), but I need to put them in a common place where those files can access these functions. 
I believe they belong to XBUtilities?

File: src/runtime_src/core/common/device.h [hunk 5]
```diff
@@ -221,6 +221,26 @@ class device : public ishim
...
+  // /**
+  //  * get_uuids() - Get UUIDs from the axlf section
+  //  *
+  //  * dtbuf: axlf section to be parsed
+  //  *
+  //  * Return: list of UUIDs
+  //  */
+  // std::vector<std::string>
+  // get_uuids(const void *dtbuf) const;
+
```
Comment (stsoe) [inline src/runtime_src/core/common/device.h:243]: Remove please

Comment (AShivangi) [reply src/runtime_src/core/common/device.h]: Sorry, I forgot that!

**Comment by AShivangi** [comment]
New sample outputs:
```
$xbutil --new validate
-----------------------------------------------------
 Invoking next generation of the xbutil application
-----------------------------------------------------
Starting validation for 1 devices

Validate device[0000:b3:00.1]
Platform            : xilinx_u200_xdma_201830_2
SC Version          : 4.2.0
Platform ID         : 0x5d1211e8

1/9 Test #1         : Kernel version
    Description     : Check if kernel version is supported by XRT
    [PASSED]
------------------------------------------------------------------
2/9 Test #2         : Aux connection
    Description     : Check if auxiliary power is connected
    Warning         : Aux power is not connected
    Warning         : Device is not stable for heavy acceleration tasks
    [PASSED WITH WARNINGS]
------------------------------------------------------------------
3/9 Test #3         : PCIE link
    Description     : Check if PCIE link is active
    [PASSED]
------------------------------------------------------------------
4/9 Test #4         : SC version
    Description     : Check if SC firmware is up-to-date
    [PASSED]
------------------------------------------------------------------
5/9 Test #5         : Verify kernel
    Description     : Run 'Hello World' kernel test
    [PASSED]
------------------------------------------------------------------
6/9 Test #6         : DMA
    Description     : Run dma test
    Details         : Host -> PCIe -> FPGA write bandwidth = 9225.017455 MB/s
    Details         : Host <- PCIe <- FPGA read bandwidth = 11706.670173 MB/s
    [PASSED]
------------------------------------------------------------------
7/9 Test #7         : Bandwidth kernel
    Description     : Run 'bandwidth kernel' and check the throughput
    [PASSED]
------------------------------------------------------------------
8/9 Test #8         : Peer to peer bar
    Description     : Run P2P test
    [SKIPPED]
------------------------------------------------------------------
9/9 Test #9         : Memory to memory DMA
    Description     : Run M2M test
    [SKIPPED]
------------------------------------------------------------------
Validation completed, but with warnings

$xbutil --new validate -qj
-----------------------------------------------------
 Invoking next generation of the xbutil application
-----------------------------------------------------
{
    "logical_devices": [
        {
            "device_id": "0000:b3:00.1",
            "platform": "xilinx_u200_xdma_201830_2",
            "sc_version": "4.2.0",
            "platform_id": "0x5d1211e8",
            "tests": [
                {
                    "id": "1",
                    "name": "Kernel version",
                    "description": "Check if kernel version is supported by XRT",
                    "status": "passed"
                },
                {
                    "id": "2",
                    "name": "Aux connection",
                    "description": "Check if auxiliary power is connected",
                    "log": [
                        {
                            "Warning": "Aux power is not connected"
                        },
                        {
                            "Warning": "Device is not stable for heavy acceleration tasks"
                        }
                    ],
                    "status": "passed"
                },
                {
                    "id": "3",
                    "name": "PCIE link",
                    "description": "Check if PCIE link is active",
                    "status": "passed"
                },
                {
                    "id": "4",
                    "name": "SC version",
                    "description": "Check if SC firmware is up-to-date",
                    "status": "passed"
                },
                {
                    "id": "5",
                    "name": "Verify kernel",
                    "description": "Run 'Hello World' kernel test",
                    "status": "passed"
                }
            ]
        }
    ]
}
```

**Comment by maxzhen** [comment]
We do not print out maximum bandwidth any more?

**Comment by AShivangi** [comment]
We should. Thanks for pointing it out!
